### ISSUE TITLE
Implementation of PPO loss

### ISSUE THREAD

**Issue by davidADSP** [nathanmargaglio/Agent-Resource-Allocation#1]
In your implementation of the PPO loss, do you not need to collapse both `prob` and `old_prob` down to a single scalar per row, instead of a vector with a single non-zero entry? Otherwise, it seems that you flood the loss with negative values as a result of the zeros if the advantage is negative.

See below for a walkthrough of what I think might be happening - I've split the implementation into a few extra variables and removed entropy for ease of explanation

```
def proximal_policy_optimization_loss(advantage, old_prediction):
    def loss(y_true, y_pred):
        prob = y_true * y_pred
        old_prob = y_true * old_prediction
        r = prob/(old_prob + 1e-10)
        clipped = K.clip(r, min_value=0.8, max_value=1.2)
        ppo_loss = K.minimum(r * advantage, clipped * advantage)

        return -K.mean(ppo_loss)
```

Let's say we have this input
```
y_true = [0, 1, 0]
y_pred = [0.3, 0.5, 0.2]
old_prediction = [0.3, 0.4, 0.3]
advantage = -1
```

Then I calculate the following:
```
prob = [0, 0.5, 0]
old_prob = [0, 0.4, 0]
r = [0, 1.25, 0]
clipped = [0.8, 1.2, 0.8]
ppo_loss = [-0.8, -1.25, -0.8]
return -0.9333
```

However, my understand of PPO loss is that only the chosen action should be used in the calculation - that is, we should collapse the vector early, so that we're working with a single scalar per row:

```
prob = K.sum(y_true * y_pred, axis = 1)
old_prob = K.sum(y_true * old_prediction, axis = 1)
``` 

This way, we get the following:

```
prob = 0.5
old_prob = 0.4
r = 1.25
clipped = 1.2
ppo_loss = -1.25
return -1.25
```

Is this correct? If not, what am I misunderstanding?

Thanks for your help!

David

**Comment by nathanmargaglio** [comment]
It's been a while since I've looked at this code, but I think your reasoning is correct.  However, I think my code agrees with your logic (granted, my code does it in a sloppy way).

> In your implementation of the PPO loss, do you not need to collapse both prob and old_prob down to a single scalar per row, instead of a vector with a single non-zero entry? Otherwise, it seems that you flood the loss with negative values as a result of the zeros if the advantage is negative.

Yes and no.  What needs to be collapsed, specifically, is the value fed into the `K.clip` function, which is `r` in this case.  This will prevent the `0` values from being clipped to `0.8`, which would cause the issue you're describing.  Of course, if you collapse both `prob` and `old_prop`, then `r` will have the correct shape.  So "yes" in that what you're suggesting should work, but "no" in that it's not the only way to achieve the same thing.

Going off your example, you have:

```
def proximal_policy_optimization_loss(advantage, old_prediction):
    def loss(y_true, y_pred):
        prob = y_true * y_pred
        old_prob = y_true * old_prediction
        r = prob/(old_prob + 1e-10)
        clipped = K.clip(r, min_value=0.8, max_value=1.2)
        ppo_loss = K.minimum(r * advantage, clipped * advantage)

        return -K.mean(ppo_loss)
```

This is very similar to the code in the repo, but it is missing one key calculation:

This line in your version:
`r = prob/(old_prob + 1e-10)`

corresponds to this line in mine:
`r = K.sum(prob/(old_prob + 1e-10), axis=1)`

Which is very similar to what you're suggesting should be done to `prob` and `old_prob`. In my code, this should result in the same values (and shape) as what you're suggesting.  If we have this input:
```
y_true = [0, 1, 0]
y_pred = [0.3, 0.5, 0.2]
old_prediction = [0.3, 0.4, 0.3]
advantage = -1
```

Then your example calculates the following:
```
prob = [0, 0.5, 0]
old_prob = [0, 0.4, 0]
r = [0, 1.25, 0]
clipped = [0.8, 1.2, 0.8]
ppo_loss = [-0.8, -1.25, -0.8]
return -0.9333
```

But my code would, instead, calculate:
```
prob = [0, 0.5, 0]
old_prob = [0, 0.4, 0]
r = 1.25
clipped = 1.2
ppo_loss = -1.25
return -1.25
```

Now, I'd like to note that this wasn't built for optimization, so your approach may be better in that sense.  But in my implementation, I'm taking advantage of the fact that `r` is necessarily a sparse vector where only one value (at most) is non-zero.  This is because the numerator of `r` is also a sparse vector.  In your version, you're utilizing this fact a few lines earlier, but the end result should be the same.

Frankly, I think your version is better.  It's a bit more clear and probably more performant since you're `prop/old_prop` is an element-wise multiplication of two vectors while mine is an element-wise multiplication of two matrices (although I wouldn't take my word for it; I'm someone writing ML code so needless, to say, I'm no expert in complexity).

**Comment by nathanmargaglio** [comment]
After looking at the repo (it's been a while) I'm seeing that you may be referring to a different function than me.

I'm looking at the loss in [PPOAgent.py](https://github.com/nathanmargaglio/Agent-Resource-Allocation/blob/master/PPOAgent.py#L82), which reflects the code I'm referencing in the above comment.

However, I'm implementing something similar in [AllocationAgent.py](https://github.com/nathanmargaglio/Agent-Resource-Allocation/blob/master/AllocationAgent.py#L64), which _doesn't_ handle that clipping in the way it should be done in PPO.  This is by design, though.  The `AllocationAgent` doesn't use a PPO loss, but a slightly modified version where the main difference is that I'm _not_ ignoring those values.  I'll spare you the explanation, but basically the allocation agent uses the entire distribution from the policy rather than just a single sample from it.

So, if there was some confusion there, I apologize.  To be honest, I'm a little disturbed that someone is actually looking at my GitHub lol.  May I ask where you found this?  And why you're looking through it?  I mean, it's public so it's not like I'm offended or anything, but this code is quite dated and has no guarantee of being functional.  I almost feel as though I should clean it up a bit just in case others are using it as reference ¯\\\_(ツ)_/¯

**Comment by davidADSP** [comment]
Thanks for the detailed answer!

Yep I completely missed the PPOAgent.py file - I was looking at AllocationAgent.py but your reasoning there makes sense. I can see now that you've implemented the sum in PPOAgent.py at a later step which is totally fine. Great that we agree.

I found your repo as I'm looking for PPO implementations using Keras backend and there are surprisingly few. There's [this one](https://github.com/LuEE-C/PPO-Keras/blob/master/Main.py) but it has the problem pointed out above - no summation over the last axis. I've seen this PPO loss function crop up in a lot of repos and I don't think it's correct, hence the reason for getting in touch. I wanted to check it wasn't my misunderstanding of how PPO works.

Thanks again for your help 👍
